intents/intent_ext.go

```diff
@@ -0,0 +1,231 @@
+
+type KeyType int
+
+const (
+	KeyTypeSECP256K1 KeyType = iota
+	KeyTypeSECP256R1
+	KeyTypeUnknown
+)
+
+const (
+	IntentNameOpenSession           = "openSession"
+	IntentNameCloseSession          = "closeSession"
+	IntentNameValidateSession       = "validateSession"
+	IntentNameFinishValidateSession = "finishValidateSession"
+	IntentNameListSessions          = "listSessions"
+	IntentNameGetSession            = "getSession"
+	IntentNameSignMessage           = "signMessage"
+	IntentNameSendTransaction       = "sendTransaction"
+)
+
+const (
+	IntentResponseCodeSessionOpened      = "sessionOpened"
+	IntentResponseCodeValidationRequired = "validationRequired"
+	IntentResponseCodeValidationStarted  = "validationStarted"
+	IntentResponseCodeValidationFinished = "validationFinished"
+	IntentResponseCodeSignedMessage      = "signedMessage"
+	IntentResponseCodeTransactionReceipt = "transactionReceipt"
+	IntentResponseCodeTransactionFailed  = "transactionFailed"
+	IntentResponseCodeGetSessionResponse = "getSessionResponse"
+)
+
+type IntentDataValidator interface {
+	IsValid() error
+}
+
+func (intent *Intent) Hash() ([]byte, error) {
+	// copy intent and remove signatures
+	var intentCopy = *intent
+	intentCopy.Signatures = nil
+
+	// Convert packet to bytes
+	packetBytes, err := canonicaljson.Marshal(intentCopy)
+	if err != nil {
+		return nil, err
+	}
+
+	// Calculate keccak256 hash
+	return crypto.Keccak256(packetBytes), nil
+}
+
+func (intent *Intent) IsValid() error {
+	if len(intent.Signatures) == 0 {
+		return fmt.Errorf("no signatures")
+	}
+
+	// check if the intent is expired
+	if intent.ExpiresAt+IntentAllowedTimeDriftInSec < uint64(time.Now().Unix()) {
+		return fmt.Errorf("intent expired")
+	}
+
+	// check if the intent is issued in the future
+	if intent.IssuedAt-IntentAllowedTimeDriftInSec > uint64(time.Now().Unix()) {
+		return fmt.Errorf("intent issued in the future")
+	}
+
+	// check if at least one signature is valid
+	if len(intent.Signers()) == 0 {
+		return fmt.Errorf("invalid signature")
+	}
+
+	// the intent is valid
+	return nil
+}
+
+func (intent *Intent) keyType(sessionId string) KeyType {
+	// handle empty session ids
+	if len(sessionId) == 0 {
+		return KeyTypeUnknown
+	}
+
+	// handle old session ids
+	if len(sessionId) == 42 {
+		return KeyTypeSECP256K1
+	}
+
+	// handle key typed session ids
+	sessionIdBytes := common.FromHex(sessionId)
+	return KeyType(sessionIdBytes[0])
+}
+
+func (intent *Intent) isValidSignature(sessionId string, signature string) error {
+	switch intent.keyType(sessionId) {
+	case KeyTypeSECP256K1:
+		return intent.isValidSignatureP256K1(sessionId, signature)
+	case KeyTypeSECP256R1:
+		return intent.isValidSignatureP256R1(sessionId, signature)
+	default:
+		return fmt.Errorf("unknown session key type")
+	}
+}
+
+// isValidSignatureP256K1 checks if the signature is valid for the given secp256k1 session
+func (intent *Intent) isValidSignatureP256K1(sessionId string, signature string) error {
```

Comment:
lets export.. ie rename to `IsValidSignatureP256K1`

Reply:
as above